systemtest/benchtest/main.go

```diff
@@ -89,12 +94,12 @@ func runBenchmark(f BenchmarkFunc) (testing.BenchmarkResult, bool, error) {
 		ok = !b.Failed()
 	})
 	if result.Extra != nil {
-		addExpvarMetrics(result, collector, *detailed)
+		addExpvarMetrics(&result, collector, *detailed)
 	}
 	return result, ok, nil
 }
 
-func addExpvarMetrics(result testing.BenchmarkResult, collector *expvar.Collector, detailed bool) {
+func addExpvarMetrics(result *testing.BenchmarkResult, collector metricsCollector, detailed bool) {
```

Comment:
Hmm, writing tests with mocking HTTP calls is too brittle. It takes away the readability and adds a lot of "extra" dependency on the test making it hard to maintain long term. Additionally, I believe for unit testing this piece of code it doesn't need to explicitly define the whole logic for how expvar is collected. So comparing a simple interface vs the above, I chose the former. Also, golang interfaces are designed to be slimmer than other OOP so there's not that much of an overhead. WDYT?

Reply:
I'd rather we introduced a more complete test for this particular piece of code rather than mock the entire surface. It will assert the correctness of the Calculations provided by `expvar.Collector`. Mocking the collector will only verify that the returned values are correct as per the mock.

I disagree that mocking using HTTP calls is too brittle and that it takes away readability. On the contrary, `httptest` provides all that's needed to mock the necessary expvar calls, and makes the collector interaction with the APM Server more obvious and easier to understand when the tests are read.

Looking at how the `expvar.Collector` works, it's shouldn't be difficult to write the necessary httptest.Server and assert its behavior in this case.

Reply:
> It will assert the correctness of the Calculations provided by expvar.Collector.

Ah, I thought we were unit testing the `addExpvarMetrics`, do we intend to test it end to end here?

Reply:
We aren't overly strict at limiting the scope and surface of the unit tests. We tend not to over-abstract components just to be able to test things in isolation as long as they are reasonably easy to mock and test together. For example https://github.com/elastic/apm-server/blob/main/model/modelindexer/indexer_test.go tests slightly more than what you'd expect strictly by the definition of "unit" tests.

Reply:
I see, I agree having integration test have their own benefits but, honestly, I am still not convinced that adding integration tests to test calculations for collector is the best approach since it can be tested via unit more simply and thoroughly. However, I see the benefit of adding tests while considering each metric since there are metric specific logic in the code. Let me update the tests.